src/server/dflycmd.cc

```diff
@@ -336,6 +339,56 @@ void DflyCmd::StartStable(CmdArgList args, ConnectionContext* cntx) {
   return rb->SendOk();
 }
 
+void DflyCmd::TakeOver(CmdArgList args, ConnectionContext* cntx) {
+  RedisReplyBuilder* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
+  string_view sync_id_str = ArgS(args, 2);
+  float timeout;
+  if (!absl::SimpleAtof(ArgS(args, 1), &timeout)) {
+    return (*cntx)->SendError(kInvalidIntErr);
+  }
+  if (timeout < 0) {
+    return (*cntx)->SendError("timeout is negative");
+  }
+
+  VLOG(1) << "Got DFLY TAKEOVER " << sync_id_str;
+
+  auto [sync_id, replica_ptr] = GetReplicaInfoOrReply(sync_id_str, rb);
+  if (!sync_id)
+    return;
+
+  unique_lock lk(replica_ptr->mu);
+  if (!CheckReplicaStateOrReply(*replica_ptr, SyncState::STABLE_SYNC, rb))
+    return;
+
+  VLOG(1) << "Takeover initiated, locking down the database.";
+  TransactionGuard tg{cntx->transaction};
+  AggregateStatus status;
+
+  auto cb = [replica_ptr = replica_ptr, timeout, &status](EngineShard* shard) {
+    FlowInfo* flow = &replica_ptr->flows[shard->shard_id()];
+
+    shard->journal()->RecordEntry(0, journal::Op::PING, 0, 0, {}, true);
+
+    absl::Time start = absl::Now();
+    while (flow->last_acked_lsn < shard->journal()->GetLsn()) {
```

Comment:
We should make sure that from the point we start initialize the TransactionGuard there are no more flows that write to journal. I believe that all the callbacks that are added to tx queue are ok, but what about expiring records? when expiring a record we replicate a DEL. Checkout the flow when calling RecordExpiry func

Reply:
It is possible to add a callback to the shard_set (not wait for it to finish) that will delete the entry and write to the journal from ExpireIfNeeded only if entry needs to be expired.
It will also solve other correctness problems we have in replication with expiring keys.
If you want to take this task, just make sure do work on it in a different PR